math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs

```diff
@@ -0,0 +1,462 @@
+
+    // c1 = a0.conjugate() + a1.conjugate() * GAMMA_12 * v + a2.conjugate() * GAMMA_14 * v^2
+    let c1 = Fp6E::new([
+        a0.conjugate(),
+        a1.conjugate() * GAMMA_12,
+        a2.conjugate() * GAMMA_14,
+    ]);
+
+    let c2 = Fp6E::new([
+        b0.conjugate() * GAMMA_11,
+        b1.conjugate() * GAMMA_13,
+        b2.conjugate() * GAMMA_15,
+    ]);
+
+    Fp12E::new([c1, c2]) //c1 + c2 * w
+}
+fn frobenius_square(
+    f: &FieldElement<Degree12ExtensionField>,
+) -> FieldElement<Degree12ExtensionField> {
+    let [a, b] = f.value();
+    let [a0, a1, a2] = a.value();
+    let [b0, b1, b2] = b.value();
+
+    let c1 = Fp6E::new([a0.clone(), GAMMA_22 * a1, GAMMA_24 * a2]);
+    let c2 = Fp6E::new([GAMMA_21 * b0, GAMMA_23 * b1, GAMMA_25 * b2]);
+
+    Fp12E::new([c1, c2])
+}
+
+////////////////// CYCLOTOMIC SUBGROUP OPERATIONS //////////////////
+/// Since the result of the Easy Part of the Final Exponentiation belongs to the cyclotomic
+/// subgroup of Fp12, we can optimize the square and pow operations used in the Hard Part.
+
+/// Computes the square of an element of a cyclotomic subgroup of Fp12.
+/// Algorithm from Constantine's cyclotomic_square_quad_over_cube
+/// https://github.com/mratsim/constantine/blob/master/constantine/math/pairings/cyclotomic_subgroups.nim#L354
+
+pub fn cyclotomic_square(a: &Fp12E) -> Fp12E {
+    // a = g + h * w
+    let [g, h] = a.value();
+    let [b0, b1, b2] = g.value();
+    let [b3, b4, b5] = h.value();
+
+    let v0 = Fp4E::new([b0.clone(), b4.clone()]).square();
+    let v1 = Fp4E::new([b3.clone(), b2.clone()]).square();
+    let v2 = Fp4E::new([b1.clone(), b5.clone()]).square();
+
+    // r = r0 + r1 * w
+    // r0 = r00 + r01 * v + r02 * v^2
+    // r1 = r10 + r11 * v + r12 * v^2
+
+    // r00 = 3v00 - 2b0
+    let mut r00 = &v0.value()[0] - b0;
+    r00 = r00.double();
+    r00 += v0.value()[0].clone();
+
+    // r01 = 3v10 -2b1
+    let mut r01 = &v1.value()[0] - b1;
+    r01 = r01.double();
+    r01 += v1.value()[0].clone();
+
+    // r11 = 3v01 - 2b4
+    let mut r11 = &v0.value()[1] + b4;
+    r11 = r11.double();
+    r11 += v0.value()[1].clone();
+    // r12 = 3v11 - 2b5
+    let mut r12 = &v1.value()[1] + b5;
+    r12 = r12.double();
+    r12 += v1.value()[1].clone();
+    // r12 = 3v11 - 2b5
+
+    let v21 = mul_fp2_by_nonresidue(&v2.value()[1]);
+    let mut r10 = &v21 + b3;
+    r10 = r10.double();
+    r10 += v21;
+
+    // 3 * ( u) * v20 - 2b3
+    let mut r02 = &v2.value()[0] - b2;
+    r02 = r02.double();
+    r02 += v2.value()[0].clone();
+
+    Fp12E::new([Fp6E::new([r00, r01, r02]), Fp6E::new([r10, r11, r12])])
+}
+
+// To understand more about how to reduce the final exponentiation
+// read "Efficient Final Exponentiation via Cyclotomic Structure for
+// Pairings over Families of Elliptic Curves" (https://eprint.iacr.org/2020/875.pdf)
+
+pub fn final_exponentiation(f: &Fp12E) -> Fp12E {
+    let f_easy_aux = f.conjugate() * f.inv().unwrap();
```

Comment:
This unwrap should be removed, perhaps pairing error could have another field `DivisionByZero` or `InvZeroError`? (Maybe converting it from the `FieldError`?)

Reply:
I agree with this.

Reply:
That is ok, but even though we are sure that `Result` will never be `Err`, handling the error is always better than using `unwrap`.

Reply:
Makes sense. Changed the `final_exponentiation` to handle the error